### PR TITLE
Clean up custom distributions

### LINKED ISSUE #5905: Add support for custom distributions deleting distributions as part of version-specific cache cleanup
As discussed in https://github.com/gradle/gradle/pull/5884#discussion_r200645330, custom distributions should also be deleted when a version has been detected as unused.

### PR DESCRIPTION
**Pull request by marcphilipp**
Custom distributions are now also deleted when a version has been
detected as unused as part of version-specific cache cleanup. In order
to determine the version of a custom distribution, the `versionNumber`
property is read from `build-receipt.properties` in
`lib/gradle-base-services-*.jar` of the distribution.

Resolves #5905.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java [hunk 1]
```diff
@@ -17,45 +17,145 @@
...
+
+import static org.apache.commons.io.filefilter.FileFilterUtils.directoryFileFilter;
+import static org.gradle.util.CollectionUtils.single;
 
 public class WrapperDistributionCleanupAction implements Action<GradleVersion> {
 
     @VisibleForTesting static final String WRAPPER_DISTRIBUTION_FILE_PATH = "wrapper/dists";
-    private static final Logger LOGGER = Logging.getLogger(WrapperDistributionCleanupAction.class);
-    private static final ImmutableList<String> DISTRIBUTION_TYPES = ImmutableList.of("bin", "all");
+    private static final Pattern DISTRIBUTION_PATTERN = Pattern.compile("^gradle-(\\d.+)-(?:(all)|(bin))$");
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java]: Do we need any pattern at all? I think we can handle all distributions using the same approach.

File: subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java [hunk 2]
```diff
@@ -17,45 +17,145 @@
...
+
+    private GradleVersion readGradleVersionFromJarFile(File jarFile) {
+        ZipFile zipFile = null;
         try {
-            deleteDistributions(version);
-        } catch (IOException e) {
-            throw new UncheckedIOException("Error deleting distribution for " + version, e);
+            zipFile = new ZipFile(jarFile);
+            return readGradleVersionFromBuildReceipt(zipFile);
+        } catch (IOException ignore) {
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java]: Some debug logging would be good here I think

File: subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java [hunk 3]
```diff
@@ -17,45 +17,145 @@
...
+    private List<File> listFiles(File baseDir, FileFilter filter) {
+        File[] dirs = baseDir.listFiles(filter);
+        return dirs == null ? Collections.<File>emptyList() : Arrays.asList(dirs);
+    }
+
+    private GradleVersion tryParseGradleVersion(String versionString) {
+        try {
+            return GradleVersion.version(versionString);
+        } catch (Exception e) {
+            return null;
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java]: Same here, some debug logging would help when deletion fails.

File: subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java (hunk 1 above)
Comment (marcphilipp) [reply subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java]: I kept it for performance reasons but it's probably a premature optimization.

Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/cache/internal/WrapperDistributionCleanupAction.java]: Yeah I think so too

**Pull request merged**